### ISSUE TITLE
VSCode plugin

### ISSUE THREAD

**Issue by bradleypeabody** [vugu/vugu#17]
See what it would take to make a basic VSCode plugin for .vugu files.  Is there a simple workaround we can just make instructions for right away in the wiki or on the doc site?

**Comment by bradleypeabody** [comment]
This actually may not be very difficult: https://code.visualstudio.com/api/language-extensions/syntax-highlight-guide, at least to get some basic stuff working.

CSS and Go code I think can just exclusively delegate to existing language behaviors (which appear to be described in the link above).

For HTML it would be nice to have some highlighting distinctions for things like `vg-if`, `vg-for` and `:`prefixed attributes, since these contain Go code expressions (and possibly could delegate to the existing go language behavior in VSCode).  But even if without that the extension would still be very useful.

**Comment by hunnble** [comment]
I'm willing to help, pls wait for a few days.

**Comment by eljoth** [comment]
Started playing around with textmate today with vugu. When it comes to syntax-highlighting the only thing left to do is preventing the html grammar from overriding the go grammar as seen in the pictures below (first without html being embedded, second vice versa).

![vugu2](https://user-images.githubusercontent.com/21229601/56003459-dcd90180-5cc6-11e9-8b10-68b35921a4eb.PNG)
![vugu1](https://user-images.githubusercontent.com/21229601/56003460-dd719800-5cc6-11e9-8380-dbc8f3c6442a.PNG)

**Comment by euphbriggs** [comment]
Have you taken a look at the Vetur plugin? Their implementation may be close to what Vugu would use.

https://github.com/vuejs/vetur/blob/master/syntaxes/vue.yaml

**Comment by bradleypeabody** [comment]
@eljoth Cool - my thought on this is the basic implementation is to just delegate to existing .html and .go file extensions wherever possible (i.e. overall it's just an HTML file, but in the `<script type="application/x-go">` section possibly that can just delegate to the go file plugin/extension).  And then on top of that some niceties can be added such as calling vugufmt and possibly some other code completion things like figuring out how `vg-if`, `vg-for`, and `vg-html` could auto-complete.  But I don't think the advanced functionality is needed for a first pass.  Just some decent syntax highlighting and calling vugufmt would be a great start.

@euphbriggs I haven't checked it out in detail, but if I end up working on this I'll have a closer look.

**Comment by eljoth** [comment]
@bradleypeabody yes, this is how I solved it. I created a new language embedding html and injecting go grammar to it into the script tags. I also added a grammar injecting keywords matching the regex `@.*=\".*\"` and `(vg-if|vg-for|vg-html)=\".*\"` so eventhandlers and vugu-keywords  - including their statement - are highlighted.

**Comment by ScreamingTaco** [comment]
@eljoth Any update on the progress or how we can help?

**Comment by eljoth** [comment]
@ScreamingTaco sorry, gut busy at university. Trying to get it back up the next time

**Comment by ScreamingTaco** [comment]
@eljoth No worries; I know what that’s like. Keep up the great work 👍

**Comment by bradleypeabody** [comment]
For reference:
https://github.com/binhonglee/vscode-vugu
https://github.com/bluetel/vscode-vugu

**Comment by bradleypeabody** [comment]
If anyone wants to work on this, I can take on updating `vugufmt` so it can just do a gofmt on the Go section and then the VSCode plugin just needs to call it.  Between this and syntax highlighting (which already works in https://github.com/binhonglee/vscode-vugu) it would immediately be pretty usable.

**Comment by binhonglee** [comment]
Hey I didn't realize I got mentioned here. As @bradleypeabody pointed out, syntax highlighting already works on the plugin I kinda hacked together a while back. [Here's the marketplace link](https://marketplace.visualstudio.com/items?itemName=binhonglee.vscode-vugu). I'll try to make time to look into further integrating it with `vugufmt` but no promises. It'll be tracked with [this issue](https://github.com/binhonglee/vscode-vugu/issues/1) opened by @bradleypeabody.

**Comment by RedCMD** [comment]
here's my VSCode extension to help with authoring TextMate grammars https://marketplace.visualstudio.com/items?itemName=RedCMD.tmlanguage-syntax-highlighter
